lib/AgreementRepository.js

```diff
@@ -0,0 +1,65 @@
+const promisify = require("es6-promisify");
+const Agreement = require('./Agreement');
+const ProfileRepository = require('./ProfileRepository');
+const MarketRepository = require('./MarketRepository');
+const TokenSingleton = require('./TokenSingleton');
+const utils = require('./web3-utils');
+
+const DEPLOY_GAS_LIMIT = 3500000;
+
+class AgreementRepository {
+
+  constructor(marketAddress, agreementJSON = require('../build/contracts/EscrowedAgreement.json')) {
+    this.AgreementContract = web3.eth.contract(agreementJSON.abi);
+    this.data = agreementJSON.unlinked_binary;
+    this.marketAddress = marketAddress;
+  }
+
+  initiateAgreement(offerAddress, quantity, transactionHashCallback) {
+    const transactionParams = {
+      from: web3.eth.accounts[0],
+      gas: DEPLOY_GAS_LIMIT,
+    }
+    return new Promise(async (resolve, reject) => {
+      try {
+        var marketContract = (await new MarketRepository().fromAddress(this.marketAddress)).marketContract;
+        var tokenAddress = await promisify(marketContract.token)();
+        var token = (await new TokenSingleton().fromAddress(tokenAddress)).contract;
+        await promisify(token.approve)(this.marketAddress, 
+              await promisify(marketContract.calculatePrice)(offerAddress, quantity),
+              { from: web3.eth.accounts[0] });
+      } catch (e) {
+        reject(e);
+      }
+      marketContract.buy(offerAddress, quantity, transactionParams,
+        async (err, tx) => {
+          if (err) {
+            reject(err);
+          } else {
+            await utils.waitForTransaction(tx, DEPLOY_GAS_LIMIT);
+            var agreementAddress = await promisify(marketContract.getNewestAgreement)();
+            resolve(new Agreement(this.AgreementContract.at(agreementAddress), this.marketAddress));
```

Comment:
Won't work if we have more than one agreement mined in the same block for one user.